WIP - Working multi-snap cluster

lazypower opened the pull request:
DO NOT MERGE!

This is an open thread to track the progress and discuss the approach to the refactoring of the etcd layer to migrate from apt delivery => snaps.

This has implications on:

- delivery mechanism
- backup/restore
- durable storage migration(s)



Initial deployment + upgrade code. Functional X node cluster. Does not survive leader termination.

Commit:
WIP - Working multi-snap cluster

Initial deployment (non-upgrade) code. Functional X node cluster. Does not survive leader termination.

lazypower commented on templates/defaults:
```diff
@@ -1,5 +1,5 @@
 # This file is rendered by Juju, manual edits will not be persisted
-ETCD_DATA_DIR=/var/lib/etcd/default
+ETCD_DATA_DIR={{ etcd_data_dir }}/{{ unit_name }}.etcd
```
This now matches what the snap is doing. I'm not 100% positive we want to keep this nomenclature but I'm for keeping it in lockstep with the snap so it stays consistent.

marcoceppi commented on reactive/etcd.py:
```diff
@@ -61,21 +60,25 @@ def check_cluster_health():
+    else:
+        subject = "peer"
+
+    status_message = "{0} with {1} known {2}.".format(unit_health, peers,
```
Could we do this instead?

```python
status_message = '{0} with {1} known peer{2}'.format(unit_health, peers, 's' if peers != 1 else '')
```

lazypower replied:
We could, I don't know what it buys us other than collapsing four lines into one, but at the sake of readability. I don't feel strongly either way however.

Commit:
Adds migration process from apt to snap

adam-stokes commented on actions/snap-upgrade:
```diff
@@ -0,0 +1,53 @@
+  else
+    # Install from the resource location, and check for correct snap.
+    MANUAL_SNAP=$(resource-get etcd)
+    snap install ${ETCD_MANUAL_SNAP} --classic
```
If you are installing a local snap you need to pass `--dangerous` as well

adam-stokes commented on actions/snap-upgrade:
```diff
@@ -0,0 +1,53 @@
+  snap refresh etcd --channel=${CONFIGURED_CHANNEL} --classic
+else
+  MANUAL_SNAP=$(resource-get etcd)
+  snap refresh ${MANUAL_SNAP}
```
Also i dont think snap refresh works for locally installed snaps

marcoceppi replied:
You'll need to just snap install /again/

Commit:
Actions snap refactor

Commit:
Tweaked peer health string

Commit:
Adds etcd3 snap support (prelim)

Tested with 3.1.2.

- Refactors template rendering to the version listed as installed in
snapd.
- Adds etcd2/3 configuration templates
- Readme updates
- Removing amulet package from test dependencies.

Commit:
Cleanup of restore and snap-upgrade work

- Move "migrate.py" work to snap-upgrade
- Refactores restore to be python based and works with both 2.x and 3.x
series.
- Changes default casing in actions.yaml bool params.

Commit:
Refactor storage to work with $SNAP_DATA

Storage was historically temp mounted, migrated, then blindly mounted
over the default $ETCD_DATA path. This now refactores the storage to
work with snaps that have access to the 'pluggable-media' interface.

This particular feature depends on
https://github.com/markshuttle/etcd-snaps/pull/4
